File: 2020-10-30-client-libraries/node/pagination.js
const Stripe = require("stripe");
const stripe = new Stripe('sk_test...');

(async () => {
  console.log("Pagination in Node!");

  // Cursor-based pagination
  // let customers = await stripe.customers.list({limit: 10})
  // const customerIds = []
  // customers.data.forEach((c) => {
  //  customerIds.push(c.id);
  // })

  // while(customers.has_more) {
  //  customers = await stripe.customers.list({
  //    limit: 10,
  //    starting_after: customers.data[customers.data.length-1].id
  //  })
  //  customers.data.forEach((c) => {
  //    customerIds.push(c.id);
  //  })
  // }

  // console.log(customerIds)
  // console.log("# of customers: %d", customerIds.length)


  // Auto-pagination
  // Documentation on different options for auto-pagination in node:
  // https://github.com/stripe/stripe-node#auto-pagination
  const customerIds = []
  for await (const customer of stripe.customers.list({limit:100})) {
    customerIds.push(customer.id)
  }

  console.log(customerIds)
  console.log("# of customers: %d", customerIds.length)
})();

File: 2020-10-30-client-libraries/go/pagination.go
package main

import (
  "fmt"
  "github.com/stripe/stripe-go/v72"
  "github.com/stripe/stripe-go/v72/customer"
)

func main() {
  stripe.Key = "sk_test..."
  fmt.Println("Pagination in Go!")

  // Cursor-based pagination
  // var customerIds []string
  // params := &stripe.CustomerListParams{}
  // params.Single = true
  // params.Filters.AddFilter("limit", "", "10")
  // i := customer.List(params)
  // for i.Next() {
  //   customer := i.Customer()
  //   customerIds = append(customerIds, customer.ID)
  // }

  // for i.Meta().HasMore {
  //   params.Filters.AddFilter("starting_after", "", i.Customer().ID)
  //   i = customer.List(params)
  //   for i.Next() {
  //     customer := i.Customer()
  //     customerIds = append(customerIds, customer.ID)
  //   }
  // }

  // fmt.Println(customerIds)
  // fmt.Printf("# of customers: %d \n", len(customerIds))

  // Auto-pagination
  var customerIds []string
  params := &stripe.CustomerListParams{}
  i := customer.List(params)
  for i.Next() {
    customer := i.Customer()
    customerIds = append(customerIds, customer.ID)
  }

  fmt.Println(customerIds)
  fmt.Printf("# of customers: %d \n", len(customerIds))
}
